kobo/apps/help/serializers.py

```diff
@@ -68,3 +76,61 @@ class Meta:
             'always_display_as_new',
         )
         read_only_fields = ('uid',)
+
+    def get_title(self, in_app_message: InAppMessage) -> str:
+        if not in_app_message.generic_related_objects:
+            return in_app_message.title
+        return self._replace_placeholders(in_app_message, in_app_message.title)
+
+    def get_snippet(self, in_app_message: InAppMessage) -> str:
+        if not in_app_message.generic_related_objects:
+            return in_app_message.snippet
+        return self._replace_placeholders(in_app_message, in_app_message.snippet)
+
+    def get_body(self, in_app_message: InAppMessage) -> str:
+        if not in_app_message.generic_related_objects:
+            return in_app_message.body
+        return self._replace_placeholders(in_app_message, in_app_message.body)
+
+    def get_html(self, in_app_message: InAppMessage) -> dict:
+        if not in_app_message.generic_related_objects:
+            return in_app_message.html
+
+        return {
+            'snippet': self._replace_placeholders(
+                in_app_message, in_app_message.html['snippet']
+            ),
+            'body': self._replace_placeholders(
+                in_app_message, in_app_message.html['body']
+            ),
+        }
+
+    def _replace_placeholders(
+        self, in_app_message: InAppMessage, value: str
+    ) -> str:
+        # This method only support transfers so far
+        transfer_identifier = (
+            f'{Transfer._meta.app_label}.{Transfer._meta.model_name}'
+        )
+        if transfer_identifier in in_app_message.generic_related_objects:
+            transfer_id = in_app_message.generic_related_objects[transfer_identifier]
+            return self._replace_placeholders_for_transfer(value, transfer_id)
+
+        return t(value)
+
+    def _replace_placeholders_for_transfer(self, value: str, transfer_id: int):
+        request = self.context['request']
+        user = get_database_user(request.user)
+        value = t(value)
+
+        try:
+            transfer = Transfer.objects.get(pk=transfer_id)
```

Comment:
:+1: glad to see it handles a lack of referential integrity, which my suggested design presents us with.